packages/auth/test/integration/flows/middleware_test_generator.ts

```diff
@@ -0,0 +1,196 @@
+/**
+ * @license
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { expect, use } from 'chai';
+import chaiAsPromised from 'chai-as-promised';
+import * as sinon from 'sinon';
+import sinonChai from 'sinon-chai';
+
+// eslint-disable-next-line import/no-extraneous-dependencies
+import {Auth, createUserWithEmailAndPassword, User} from '@firebase/auth';
+import { randomEmail } from '../../helpers/integration/helpers';
+
+use(chaiAsPromised);
+use(sinonChai);
+
+export function generateMiddlewareTests(authGetter: () => Auth, signIn: () => Promise<unknown>): void {
```

Comment:
@jamesdaniels I tried to be exhaustive with these tests. Please take a look and see if I'm missing any big use case.

Cross-tab testing is done down in the webdriver/* files.

I'll add the onAbort callback to the tests once that's merged